### PR TITLE
Run CRAN check with `r-devel` when releasing a new version & daily

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

- Run the CRAN check with `r-devel` when releasing a new version.
- Run the CRAN check with `r-devel` daily to catch issues as early as possible.

## How is this patch tested?

By manually confirming `r-devel` is used on a release branch, and not used on a non-release branch.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -106,6 +107,9 @@ jobs:
         Rscript -e 'install.packages("h2o", repos="http://cran.r-project.org")'
     - name: Build package
       run: |
+        # Use r-devel on a release branch to run the same submission checks as CRAN.
```
Comment (dbczumar) [inline .github/workflows/master.yml:110]: If we need `rocker` in order to test development versions of R, then we need to update `r-devel` to use `rocker`. However, our internal CI is incompatible with `rocker`, so we would then have to update internal CI to use another image. But, `Dockerfile.dev` also uses `rocker`, so we'd need to change `Dockerfile.dev` to use `r-base`, right?

Comment (harupy) [reply .github/workflows/master.yml:110]: @dbczumar Thanks for the comment!

> our internal CI is incompatible with rocker.

Yep, can we do the following?

- In our internal CI, use `rocker:4.2.0` (which uses [ubuntu 20.04](https://github.com/rocker-org/rocker-versioned2/blob/master/dockerfiles/r-ver_4.2.0.Dockerfile)) and just build the package without running submission checks.
- In GitHub Actions, run submission checks using `rocker:r-devel`.

Comment (harupy) [reply .github/workflows/master.yml:110]: Will `rocker/r-ver:4.2.0` switch to ubuntu 20.04 in the future? https://github.com/rocker-org/rocker-versioned2/issues/282#issuecomment-1108389687 says:

> Base images for R versions released before 2022-07-19 (i.e. 4.1.3, 4.2.0) will remain ubuntu:focal and will not switch.

Comment (harupy) [reply .github/workflows/master.yml:110]: @eitsupi Thanks for the comment! Is there any other image than `rocker/r-ver:devel` that provides a development version of R?

File: .github/workflows/r.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    branches:
+      - master
+      - branch-[0-9]+.[0-9]+
+  pull_request:
+    branches:
+      - master
+      - branch-[0-9]+.[0-9]+
+  schedule:
+    # Run this workflow daily at 7:00 UTC
+    - cron: "0 7 * * *"
```
Comment (harupy) [inline .github/workflows/r.yml:14]: I'll add a badge on README later in a follow-up PR.

File: .github/workflows/r.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      run: |
+        if [ "$GITHUB_EVENT_NAME" = "schedule" ]; then
+          USE_R_DEVEL=true
+        elif [ "$GITHUB_EVENT_NAME" = "pull_request" ]; then
+          # Use r-devel on a pull request targeted to a release branch
+          USE_R_DEVEL=$([[ $GITHUB_BASE_REF =~ branch-[0-9]+\.[0-9]+$ ]] && echo true || echo false)
+        else
+          # Use r-devel on a push to a release branch
+          USE_R_DEVEL=$([[ $GITHUB_REF_NAME =~ branch-[0-9]+\.[0-9]+$ ]] && echo true || echo false)
+        fi
```
Comment (harupy) [inline .github/workflows/r.yml:68]: Uses `r-devel` when releasing a new version.

Comment (harupy) [reply .github/workflows/r.yml:68]: Do we still need a manual confirmation stage for the CRAN check?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

File: .github/workflows/master.yml (hunk 1 above)
Comment (eitsupi) [reply .github/workflows/master.yml]: The Rocker project have `rocker/r-devel` and `rocker/drd`.
https://github.com/rocker-org/r-devel
https://github.com/rocker-org/drd

However, these are only built once a week, so if you want to try a newer version, `rocker/r-ver:devel`, which is built daily, is a good choice, I think.

**Pull request merged**